src/template.c

```diff
@@ -7302,6 +7244,72 @@ bool TemplateInstance::hasNestedArgs(Objects *args, bool isstatic)
     return nested != 0;
 }
 
+/*****************************************
+ * Append 'this' to the enclosing module members[] so the semantic routines
+ * will get called on the instance members.
+ */
+Dsymbols *TemplateInstance::appendToModuleMember()
+{
+    Module *mi = minst;
+    if (mi)
+    {
+        if (mi->isRoot())
+        {
+            /* If both modules mi and md where the template is declared are
+             * roots, instead use md to store the generated code in it.
+             */
+            Module *md = tempdecl->getModule();
+            if (md && md->isRoot())
+                mi = md;
```

Comment:
> That seems to be a new rule, why is it needed?

By the `mi` adjustment, when a template `X` is declared in module `m.d`, and `m.d` is actually listed in command line, its instance will be constantly stored in `m.obj`.

> Who actually introduced that whacky minst concept, there can be multiple modules instantiating a template in D?

Walter had introduced the concept in #2550 (Later I renamed [instantiatingModule](https://github.com/D-Programming-Language/dmd/pull/2550/files#diff-46c6a9dc108e93f73a527f235351cfc4R68) to `minst`).
With a chain of template instantiations (A!() instantiates B!() instantiates C!() ...), when the first instance `A!()` exists in non-root module, all of their code generation can be elided. If `minst->isRoot() == false`, it means the instance is a part of non-root instantiation chain.